wallet/controllers.go

```diff
@@ -0,0 +1,149 @@
+	"crypto"
+	"database/sql"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/brave-intl/bat-go/middleware"
+	"github.com/brave-intl/bat-go/utils/altcurrency"
+	appctx "github.com/brave-intl/bat-go/utils/context"
+	errorutils "github.com/brave-intl/bat-go/utils/errors"
+	"github.com/brave-intl/bat-go/utils/handlers"
+	"github.com/brave-intl/bat-go/utils/httpsignature"
+	"github.com/brave-intl/bat-go/utils/logging"
+	walletutils "github.com/brave-intl/bat-go/utils/wallet"
+	"github.com/go-chi/chi"
+	uuid "github.com/satori/go.uuid"
+)
+
+// Router for suggestions endpoints
+func Router(service *Service) chi.Router {
+	r := chi.NewRouter()
+	r.Method("GET", "/{paymentId}", middleware.InstrumentHandler("GetWallet", GetWallet(service)))
+	return r
+}
+
+// LookupPublicKey based on the HTTP signing keyID, which in our case is the walletID
+func (service *Service) LookupPublicKey(ctx context.Context, keyID string) (*httpsignature.Verifier, error) {
+	walletID, err := uuid.FromString(keyID)
+	if err != nil {
+		return nil, errorutils.Wrap(err, "KeyID format is invalid")
+	}
+
+	wallet, err := service.GetWallet(walletID)
+	if err != nil {
+		return nil, errorutils.Wrap(err, "error getting wallet")
+	}
+
+	if wallet == nil {
+		return nil, nil
+	}
+
+	var publicKey httpsignature.Ed25519PubKey
+	if len(wallet.PublicKey) > 0 {
+		var err error
+		publicKey, err = hex.DecodeString(wallet.PublicKey)
+		if err != nil {
+			return nil, err
+		}
+	}
+	tmp := httpsignature.Verifier(publicKey)
+	return &tmp, nil
+}
+
+// LinkWalletRequest holds the data necessary to update a wallet with an anonymous address
+type LinkWalletRequest struct {
+	SignedTx         string     `json:"signedTx"`
+	AnonymousAddress *uuid.UUID `json:"anonymousAddress"`
+}
+
+// GetWalletResponse gets wallet info tied to a wallet id
+type GetWalletResponse struct {
+	Wallet *walletutils.Info `json:"wallet"`
+}
+
+func validateHTTPSignature(ctx context.Context, r *http.Request, signature string) (string, error) {
+	// validate that the signature in the header is valid based on public key provided
+	var s httpsignature.Signature
+	err := s.UnmarshalText([]byte(signature))
+	if err != nil {
+		return "", fmt.Errorf("invalid signature: %w", err)
+	}
+
+	// Override algorithm and headers to those we want to enforce
+	s.Algorithm = httpsignature.ED25519
+	s.Headers = []string{"digest", "(request-target)"}
+	var publicKey httpsignature.Ed25519PubKey
+	if len(s.KeyID) > 0 {
+		var err error
+		publicKey, err = hex.DecodeString(s.KeyID)
+		if err != nil {
+			return "", fmt.Errorf("failed to hex decode public key: %w", err)
+		}
+	} else {
+		// there was no KeyId in the Signature
+		if err != nil {
+			return "", errors.New("no KeyId found in the HTTP Signature")
+		}
+	}
+	pubKey := httpsignature.Verifier(publicKey)
+	if err != nil {
+		return "", err
+	}
+	if pubKey == nil {
+		return "", errors.New("invalid public key")
+	}
+
+	valid, err := s.Verify(pubKey, crypto.Hash(0), r)
```

Comment:
does crypto.Hash(0) do anything here? IIUC, it eventually gets passed as the `opts` to https://github.com/brave-intl/bat-go/blob/master/utils/httpsignature/ed25519.go#L18, which is an unused variable?

Reply:
This is the way of saying what hashing mechanism we want.  There are a list of constants (iota based in crypto library) and we are casting the number 0 to a crypto.Hash type.  Effectively this is the same as saying use the crypto.Hash typed constant with the value 0.

Reply:
Basically casting the number 0 into the Hash type constant in the crypto library, and passing that to verify... Verify performs a hash of that kind, on the request, then uses the pubKey to verify it.

Reply:
@husobee thanks. i see where this gets used for ed25519 signing (https://golang.org/pkg/crypto/ed25519/#PrivateKey.Sign) but i don't see it in Verify: https://github.com/brave-intl/bat-go/blob/master/utils/httpsignature/ed25519.go#L23. 

(ed25519 can't handle pre-hashed messages, so crypto.Hash(0) indicates no hashing. i guess the Sign interface expects this to be passed explicitly, whereas Verify doesn't.)

Reply:
OH, I misunderstood.  I see what you mean, in that Ed25519PubKey does not use the SignerOpts in the Verify implementation.  My best guess as to why this is setup this way would be to make it easy for other Verify-ing mechansims to pass in Signer options if they are needed for that particular scheme?  By using a Verify interface it makes it easier for us to support other mechanisms in the future that might require these options.  Although we only support one right now, this might be considered premature optimization.